HW0/P4/P4.py

```diff
@@ -24,53 +24,68 @@
     # Create 3D axes for plotting
     ax = Axes3D(plt.figure())
 
-    #####################
+
     # Part 1:
-    #
-    # Load true trajectory and plot it
-    # Normally, this data wouldn't be available in the real world
-    #####################
+    s_true=np.loadtxt('P4_trajectory.txt',delimiter=',')
+    true_x=s_true[:,0]
+    true_y=s_true[:,1]
+    true_z=s_true[:,2]
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '--b', label='True trajectory')
+    ax.plot(true_x, true_y, true_z,'--b', label='True trajectory')
 
-    #####################
+    
     # Part 2:
-    #
-    # Read the observation array and plot it (Part 2)
-    #####################
-
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '.g', label='Observed trajectory')
+    s_measure=np.loadtxt('P4_measurements.txt',delimiter=',')
+    measure_x=s_measure[:,0]/rx
+    measure_y=s_measure[:,1]/ry
+    measure_z=s_measure[:,2]/rz
+    
+    ax.plot(measure_x, measure_y, measure_z,'.g', label='Observed trajectory')
 
-    #####################
     # Part 3:
-    # Use the initial conditions and propagation matrix for prediction
-    #####################
-
-    # A = ?
-    # a = ?
-    # s = ?
+    A=np.array([[1,0,0,dt,0,0],[0,1,0,0,dt,0],[0,0,1,0,0,dt],[0,0,0,1-c*dt,0,0],[0,0,0,0,1-c*dt,0],[0,0,0,0,0,1-c*dt]])
+    a=np.array([[0],[0],[0],[0],[0],[g*dt]])
+    S=np.zeros((6,K))
 
     # Initial conditions for s0
+    S[:,0]=list([0,0,2,15,3.5,4])
+
     # Compute the rest of sk using Eq (1)
+    i=0
+    while i<K-1:
+        S[:,i+1]=(np.dot(A,S[:,i].reshape((6,1)))+a).reshape(6)
+        i+=1
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-k', label='Blind trajectory')
+    px=S[0,:].reshape(K)
+    py=S[1,:].reshape(K)
+    pz=S[2,:].reshape(K)
+   
+    ax.plot(px,py,pz,'-k', label='Blind trajectory')
 
-    #####################
-    # Part 4:
-    # Use the Kalman filter for prediction
-    #####################
 
-    # B = ?
-    # C = ?
+    # Part 4:
+    B=np.diag((bx,by,bz,bvx,bvy,bvz))
+    C=np.concatenate((np.diag((rx,ry,rz)),np.zeros((3,3))))
 
     # Initial conditions for s0 and Sigma0
-    # Compute the rest of sk using Eqs (2), (3), (4), and (5)
+    Sigma=0.01*np.matrix(np.identity(6))
+    s=np.zeros((6,K))
+    s[:,0]=list([0,0,2,15,3.5,4])
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-r', label='Filtered trajectory')
+    # Compute the rest of sk using Eqs (2), (3), (4), and (5)
+    j=0
+    while j<K-1:
+    	s_predict=np.dot(A,s[:,j].reshape((6,1)))+a
```

Comment:
The problem states do define four python functions, namely the predictS, predictSig, update Sig, and updateS. So rather than putting in the loop as you did, it will be cleaner if you define separate functions outside.